server/server.go

```diff
@@ -267,3 +278,44 @@ func WithServerConfig(config Config) Option {
 		s.baseCtx = context.WithValue(s.baseCtx, ctxConfigKey{}, config)
 	}
 }
+
+// WithTracer is an option function that initializes an opentelemtry tracer for the server requests.
+// It takes the following parameter:
+//
+//	an `exporter` function argument. This function is used to set the otel exporter of your choice for tracing.
+//	Please refer to https://opentelemetry.io/docs/languages/go/exporters/ to know choices.
+//
+// Adding a tracer option is a must if you want to create a span for your requests.
+func WithTracer(exporter func(ctx context.Context) (sdktrace.SpanExporter, error)) Option {
+	return func(s *Server) {
+		exp, err := exporter(s.baseCtx)
+		if err != nil {
+			panic(fmt.Sprintf("failed to initialize exporter: %v", err))
```

Comment:
@shan-96 Middleware looks good, but I don't think we need to be aware of open telemetry at server abstraction. Do we? it's out of scope of responsibility of it.

Reply:
I'm equating tracing as a feature on the same level as profiling. And that's why added on server level. But there is one more reason. consider a server is has multiple channels (say 3 and 2 of them use telemetry). Now who will set 
```go
otel.SetTracerProvider(s.TracerProvider)
```
because from the docs...
```go
// SetTracerProvider registers `tp` as the global trace provider.
func SetTracerProvider(tp trace.TracerProvider) {
	global.SetTracerProvider(tp)
}
```
this must be done exactly once per application and should not be overridden by any channel

Reply:
@shan-96 I see your point but it's not exactly the same, main benefit of having profiler option that we're using http server of wasabi. 

Here we have only attachment tracer to live cycle of wasabi server which can be solved with simple defer function.

I think we should not bring it to server. configuration of Otel should not be our concern, help with code instrumentation should be

Reply:
and probably, you have a point, bringing profiler to Server wasn't too bright decision :)

Reply:
ok but what about the trace provider being set on application level?

and maybe there is an architectural decision to be taken here if we want things like profiler and tracing to be defined by server or middleware.

I agree what sort of observability the server can have might be defined by its backend, but then then the server should (or can? :D ) be the one to specify what is sort of observables are enabled?

Reply:
@ksysoev do you agree to keep the trace provider logic on channel level instead?

Reply:
There are two problems why I want to do that. First is that the provider must be shutdown at some point. This cannot be on the middleware level. It doesnt make sense to build the provider and shutdown on every websocket request. We can do it per channel and control this when the channel closes down

Reply:
@shan-96 I think there is a confusion, wasabi is not full cycle application, it's a toolkit that can be used for creating ws application. we don't need to build-in lifecycle management of the other libraries, it's way more flexing to keep it outside of server abstraction like this:

```golang
main () {
t := sdktrace.NewTracerProvider(
 	sdktrace.WithBatcher(exp),
 	sdktrace.WithResource(r),
)

defer t.Shutdown()


// here we create middleware and pass tracer

server.NewServer()
serve.Run()
}
```